### DISCUSSION TITLE
Ability to use VideoTexture for Android video players - Write to Surface before rendered in AIR?

### DISCUSSION THREAD

**Discussion by teotigraphix** [airsdk/Adobe-Runtime-Support#2635] [Q&A] [answered]
Before I go down a rabbit hole in my mind, @ajwfrost is it possible for me to draw on VideoTexture in native code?

Use case, I have an audio system natively, could I make draws calls on that VideoTexture  surface natively before it's feed into the runtime to make a low level graphic visualizer?

I have not researched this, so I am basically dreaming here but is that essentially what "ExoPlayer" is doing?

**Comment by ajwfrost** [comment, marked as answer]
Hi

Potentially, yes, but I think you'd need to be doing some special low level stuff to make that work... it's designed to fit in with a media pipeline, as far as I understand it, so we'd exposed access to the `SurfaceTexture` so that this can be passed in for ExoPlayer etc..
https://developer.android.com/reference/android/graphics/SurfaceTexture

So if you're able to create a component that can do the same thing, then it should be possible to make this work - just get the SurfaceTexture from the AS3 VideoTexture object and tell it what the dimensions are...

We had still also wondered about injection of display objects i.e. rendering a custom DisplayObject using an external renderer, but there are so many conditions here i.e. ensuring that you use the appropriate platform-specific mechanism based on renderMode etc... but hopefully at some point, that would provide a similar/simpler option!

thanks

**Reply by teotigraphix** [reply]
> provide a similar/simpler option

That's great to hear. I think the more you could integrate these finer points in the runtime, it just becomes more robust and easier to talk about with devs interested in AIR's newer capabilities.

As far as ExoPlayer, that was just for an example of an Android "component" being used to render the frame and then pass to VideoTexture.

I had used this years ago, like 2011 or so, I did read the API docs on Surface/SurfaceView/SurfaceHolder so I get the main gist now. If I get a working version of an ANE I will post it here. (months probably but on the list)

Thanks
